### PR TITLE
GFORMS-1991: Allow for ATL data to be presented as bulletedList or nu…

### PR DESCRIPTION
**Pull request by lxol**
…mberedList in info fields

fix unescaped user data in the startstrings

### CODE REVIEW

File: app/uk/gov/hmrc/gform/gform/ExprUpdater.scala [hunk 1]
```diff
@@ -62,6 +62,8 @@ class ExprUpdater(index: Int, baseIds: List[FormComponentId]) {
     case PeriodExt(_, _)                      => expr
     case IndexOf(_, _)                        => expr // This is not explanded on purpose, so it can be used correctly inside ATL
     case RemoveSpaces(formComponentId)        => RemoveSpaces(expandFcId(formComponentId))
+    case NumberedList(formComponentId)        => RemoveSpaces(expandFcId(formComponentId))
+    case BulletedList(formComponentId)        => RemoveSpaces(expandFcId(formComponentId))
```
Comment (DursunGundogan) [inline app/uk/gov/hmrc/gform/gform/ExprUpdater.scala:66]:
```suggestion
    case NumberedList(formComponentId)        => NumberedList(expandFcId(formComponentId))
    case BulletedList(formComponentId)            => BulletedList(expandFcId(formComponentId))
```

File: app/uk/gov/hmrc/gform/sharedmodel/graph/DependencyGraph.scala [hunk 2]
```diff
@@ -185,7 +185,28 @@ object DependencyGraph {
...
+              GraphNode.Expr(FormCtx(fc.id)) ~> GraphNode.Simple(fc.id)
+            )
+          )
+    val govukListNodes = graph.nodes.flatMap { n =>
+      n.value match {
+        case g @ GraphNode.Expr(NumberedList(fcId)) => govukListEdges(g, fcId)
+        case g @ GraphNode.Expr(BulletedList(fcId)) => govukListEdges(g, fcId)
+        case e                                      => List.empty[DiEdge[GraphNode]]
+      }
+    }.toSet
```
Comment (josef-vlach) [inline app/uk/gov/hmrc/gform/sharedmodel/graph/DependencyGraph.scala:207]: Why is this here?

**Review by josef-vlach — CHANGES_REQUESTED** [review]
Referring ATL `addressId` outside ATL should work, but it doesn't. First step should be to fix this problem (I've already mentioned it last time) and then build this functionality on top of that solution, otherwise this PR is introducing lots of unnecessary complexity.

File: app/uk/gov/hmrc/gform/sharedmodel/graph/DependencyGraph.scala (hunk 2 above)
Comment (lxol) [reply app/uk/gov/hmrc/gform/sharedmodel/graph/DependencyGraph.scala:207]: alternatively we can evaluate `FormCtx(addressid)` to AddressResult(addresslines)` or `ListResult(List(AddressResult))` but we will need to review and update all the usages as they all expect `AddressResult(addresslines)`

Comment (josef-vlach) [reply app/uk/gov/hmrc/gform/sharedmodel/graph/DependencyGraph.scala:207]: > For `addressId` that refers to ATL address we need `ListResult(AddressResult(...))`

And this is exactly what we need to fix. [This should return]( https://github.com/hmrc/gform-frontend/blob/46f798b67926b82957b1e6ec4aefe5ecf535429a/app/uk/gov/hmrc/gform/eval/EvaluationResults.scala#L355) `ListResult(AddressResult(...))` if `addressId` is ATL component.

File: app/uk/gov/hmrc/gform/eval/ExpressionResult.scala [hunk 3]
```diff
@@ -333,6 +333,13 @@ sealed trait ExpressionResult extends Product with Serializable {
       None
     )(_ => None)(_ => None)(_ => None)
 
+  def govukListRepresentation(typeInfo: TypeInfo, messages: Messages): List[String] =
```
Comment (josef-vlach) [inline app/uk/gov/hmrc/gform/eval/ExpressionResult.scala:336]: Let call this method only `listRepresentation`

File: app/uk/gov/hmrc/gform/eval/smartstring/SmartStringEvaluatorFactory.scala [hunk 4]
```diff
@@ -152,5 +163,29 @@ class RealSmartStringEvaluatorFactory() extends SmartStringEvaluatorFactory {
           .evalAndApplyTypeInfo(typeInfo)
           .optionRepresentation
           .fold(stringRepresentation(typeInfo))(f(_))
+
+      private def govukListRepresentation(typeInfo: TypeInfo, isBulleted: Boolean, markDown: Boolean): String = {
+        formModelVisibilityOptics.evalAndApplyTypeInfo(typeInfo)
```
Comment (josef-vlach) [inline app/uk/gov/hmrc/gform/eval/smartstring/SmartStringEvaluatorFactory.scala:168]: Stray expression

File: app/uk/gov/hmrc/gform/sharedmodel/VariadicFormData.scala [hunk 5]
```diff
@@ -118,6 +119,13 @@ case class VariadicFormData[S <: SourceOrigin](data: Map[ModelComponentId, Varia
       }
     }
 
+  def distinctIndexedComponentIds(modelComponentId: ModelComponentId): List[IndexedComponentId] =
+    forBaseComponentIdLessThen(modelComponentId)
+      .map(_._1.fold(p => p.indexedComponentId)(i => i.indexedComponentId))
```
Comment (josef-vlach) [inline app/uk/gov/hmrc/gform/sharedmodel/VariadicFormData.scala:124]:
```suggestion
   forBaseComponentId(modelComponentId.indexedComponentId.baseComponentId)
      .map { case (modelComponentId, _) => modelComponentId.fold(p => p.indexedComponentId)(i => i.indexedComponentId) }
```
`forBaseComponentIdLessThen` should not be used here, we are not in ATL. Also let's get rid of `_._1`

File: app/uk/gov/hmrc/gform/eval/EvaluationResults.scala [hunk 6]
```diff
@@ -343,16 +359,21 @@ case class EvaluationResults(
...
-
-          val variadicValues: List[Option[VariadicValue]] = addressAtoms.map(atom => recData.variadicFormData.get(atom))
-          val addressLines = variadicValues.collect { case Some(VariadicValue.One(value)) if value.nonEmpty => value }
-          ExpressionResult.AddressResult(addressLines)
+          val modelComponentId = formComponentId.modelComponentId
+          val isModelFormComponentIdPure = modelComponentId.indexedComponentId.isPure
+          val isReferenceIndexed = evaluationContext.indexedComponentIds.exists(
+            _.baseComponentId === modelComponentId.baseComponentId
+          )
+          if (isModelFormComponentIdPure && isReferenceIndexed) {
```
Comment (josef-vlach) [inline app/uk/gov/hmrc/gform/eval/EvaluationResults.scala:367]: Now we are duplicating this logic in two places, can we remove the duplication?

**Review by josef-vlach — CHANGES_REQUESTED** [review]
Couple of issues to resolve, see the comments.

But also two observations:

1. if `agentFullName` is used only as `${bulletedList(agentFullName)}` list is not being produced
2. `${addressId}` expression used outside ATL do not produce any output. (`def addressRepresentation(typeInfo: TypeInfo)`  need to handle possibility of a `List`

File: app/uk/gov/hmrc/gform/eval/EvaluationResults.scala (hunk 6 above)
Comment (lxol) [reply app/uk/gov/hmrc/gform/eval/EvaluationResults.scala:367]: >. (def addressRepresentation(typeInfo: TypeInfo) need to handle possibility of a List
but it returns `List[String]` where each string is part of the address. If this method takes ListResult(List(AddressResult)) the return type of List[String] wouldn't make sense as it will contain all lines for all addresses in one List
Unless I misunderstood that.

Comment (lxol) [reply app/uk/gov/hmrc/gform/eval/EvaluationResults.scala:367]: >if agentFullName is used only as ${bulletedList(agentFullName)} list is not being produced
didn't get that

Comment (lxol) [reply app/uk/gov/hmrc/gform/eval/EvaluationResults.scala]: > if agentFullName is used only as ${bulletedList(agentFullName)} list is not being produced

do you mean that if `agentFulName` in the smart string without `bulletedList` then it is rendered as empty? I don't know how to render it tbh. Do we have any requirements on how to render list of addresses to a string?

Comment (josef-vlach) [reply app/uk/gov/hmrc/gform/eval/EvaluationResults.scala]: > but it returns List[String] where each string is part of the address. If this method takes ListResult(List(AddressResult)) the return type of List[String] wouldn't make sense as it will contain all lines for all addresses in one List
Unless I misunderstood that.

If `def addressRepresentation(...): List[String` is called on a `ListResult` of addresses we resulting list should contains individual addresses concatenated by `,<space> `. It is better default than showing nothing.

**Review by josef-vlach — APPROVED** [review]
